packages/flutter/lib/src/painting/borders.dart

```diff
@@ -259,6 +259,318 @@ class BorderSide {
+  String toString() {
+    return '$runtimeType()';
+  }
+}
+
+/// Represents the addition of two otherwise-incompatible borders.
+///
+/// The borders are listed from the outside to the inside.
+class _CompoundBorder extends ShapeBorder {
+  _CompoundBorder(this.borders) {
+    assert(borders != null);
+    assert(borders.length >= 2);
+    assert(!borders.any((ShapeBorder border) => border is _CompoundBorder));
+  }
+
+  final List<ShapeBorder> borders;
+
+  @override
+  EdgeInsetsGeometry get dimensions {
+    return borders.fold<EdgeInsetsGeometry>(
+      EdgeInsets.zero,
+      (EdgeInsetsGeometry previousValue, ShapeBorder border) {
+        return previousValue.add(border.dimensions);
+      },
+    );
+  }
+
+  @override
+  ShapeBorder add(ShapeBorder other, { bool reversed: false }) {
+    // This wraps the list of borders with "other", or, if "reversed" is true,
+    // wraps "other" with the list of borders.
+    // If "reversed" is false, "other" should end up being at the start of the
+    // list, otherwise, if "reversed" is true, it should end up at the end.
+    // First see if we can merge the new adjacent borders.
+    if (other is! _CompoundBorder) {
+      // Here, "ours" is the border at the side where we're adding the new
+      // border, and "merged" is the result of attempting to merge it with the
+      // new border. If it's null, it couldn't be merged.
+      final ShapeBorder ours = reversed ? borders.last : borders.first;
+      final ShapeBorder merged = ours.add(other, reversed: reversed)
+                             ?? other.add(ours, reversed: !reversed);
+      if (merged != null) {
+        final List<ShapeBorder> result = <ShapeBorder>[];
+        result.addAll(borders);
+        result[reversed ? result.length - 1 : 0] = merged;
+        return new _CompoundBorder(result);
+      }
+    }
+    // We can't, so fall back to just adding the new border to the list.
+    final List<ShapeBorder> mergedBorders = <ShapeBorder>[];
+    if (reversed)
+      mergedBorders.addAll(borders);
+    if (other is _CompoundBorder) {
+      mergedBorders.addAll(other.borders);
+    } else {
+      mergedBorders.add(other);
+    }
+    if (!reversed)
+      mergedBorders.addAll(borders);
+    return new _CompoundBorder(mergedBorders);
+  }
+
+  @override
+  ShapeBorder scale(double t) {
+    return new _CompoundBorder(
+      borders.map<ShapeBorder>((ShapeBorder border) => border.scale(t)).toList()
+    );
+  }
+
+  @override
+  ShapeBorder lerpFrom(ShapeBorder a, double t) {
+    return _CompoundBorder.lerp(a, this, t);
+  }
+
+  @override
+  ShapeBorder lerpTo(ShapeBorder b, double t) {
+    return _CompoundBorder.lerp(this, b, t);
+  }
+
+  static _CompoundBorder lerp(ShapeBorder a, ShapeBorder b, double t) {
+    assert(a is _CompoundBorder || b is _CompoundBorder); // Not really necessary, but all call sites currently intend this.
+    final List<ShapeBorder> as = a is _CompoundBorder ? a.borders : <ShapeBorder>[a];
```

Comment:
While `as` is a keyword, it's technically not a reserved word, so not technically a real issue. My first thought was `ab` and `bb` but now I've mixed `a`s and `b`s, so unless you've got a better idea, I suppose leave as-is. :)

Reply:
wow, `as` not being reserved is weird.

Reply:
changed to aList and bList